frame/support/src/migrations.rs

```diff
@@ -15,19 +15,128 @@
+	traits::{
+		GetStorageVersion, NoStorageVersionSet, OnRuntimeUpgrade, PalletInfoAccess, StorageVersion,
+	},
 	weights::{RuntimeDbWeight, Weight},
 };
 use impl_trait_for_tuples::impl_for_tuples;
 use sp_core::Get;
 use sp_io::{hashing::twox_128, storage::clear_prefix, KillStorageResult};
 use sp_std::marker::PhantomData;
+
 #[cfg(feature = "try-runtime")]
 use sp_std::vec::Vec;
 
+/// Make it easier to write versioned runtime upgrades.
+///
+/// VersionedRuntimeUpgrade allows developers to write migrations without worrying about checking
+/// and setting storage versions. Instead, the developer wraps their migration in this struct which
+/// takes care of version handling using best practices.
+///
+/// It takes 5 type parameters:
+/// - `From`: The version being upgraded from.
+/// - `To`: The version being upgraded to.
+/// - `Inner`: An implementation of `OnRuntimeUpgrade`.
+/// - `Pallet`: The Pallet being upgraded.
+/// - `Weight`: The runtime's RuntimeDbWeight implementation.
+///
+/// When a VersionedRuntimeUpgrades `on_runtime_upgrade` method is called, the on-chain version of
+/// the pallet is compared to `From`. If they match, `Inner::on_runtime_upgrade` is called and the
+/// pallets on-chain version is set to `To`. Otherwise, a warning is logged notifying the developer
+/// that the upgrade was a noop and should probably be removed.
+///
+/// Example:
+/// ```ignore
+/// // Migrations to pass to the Executive pallet.
+/// pub type Migrations = (
+/// 	// ...other migrations
+/// 	VersionedRuntimeUpgrade<
+/// 		Const16<4>, // From on-chain version 4
+/// 		Const16<5>, // To on-chain version 5
+/// 		parachains_configuration::migration::v5::MigrateToV5<Runtime>,
+/// 		Configuration,
+/// 		RocksDbWeight,
+/// 	>,
+/// 	VersionedRuntimeUpgrade<
+/// 		Const16<5>, // From on-chain version 5
+/// 		Const16<7>, // To on-chain version 7
+/// 		parachains_configuration::migration::v6::MigrateToV6<Runtime>,
+/// 		Configuration,
+/// 		RocksDbWeight,
+/// 	>,
+/// 	// ...other migrations
+/// );
+/// ```
+pub struct VersionedRuntimeUpgrade<From, To, Inner, Pallet, Weight> {
+	_marker: PhantomData<(From, To, Inner, Pallet, Weight)>,
+}
+
+/// Implementation of the `OnRuntimeUpgrade` trait for `VersionedRuntimeUpgrade`.
+///
+/// Primarily, it
+/// 1. Logs a warning in `pre_upgrade` if the storage version specified in the pallet appears to be
+/// misconfigured before returning `Inner::pre_upgrade`.
+///
+/// 2. Performs the actual runtime upgrade in `on_runtime_upgrade` only if the on-chain version of
+/// the pallets storage matches `From`, and after the upgrade sets the on-chian storage to `To`. If
+/// the versions do not match, it writes a log notifying the developer that the migration is a noop.
+impl<
+		From: Get<u16>,
+		To: Get<u16>,
+		Inner: OnRuntimeUpgrade,
+		Pallet: GetStorageVersion<CurrentStorageVersion = StorageVersion> + PalletInfoAccess,
+		DbWeight: Get<RuntimeDbWeight>,
```

Comment:
For readability purposes, I would consider moving `Pallet` and `DbWeight` to be associated types rather than generics and hardcoded. This might require you to add another trait. When a pallet is implementing this new trait, it will fix its `Pallet` and `DbWeight` (both of which can be obtained from `system`).

Then, in the runtime, when you are wrapping things in `VersionedRuntimeUpgrade`, you only see `(VersionedRuntimeUpgrade<1, 2, Foo>, VersionedRuntimeUpgrade<5, 6, Bar>)`. 

Possibly not worth it, especially as I am still critical of the fundamental approach.

Reply:
I'm able to remove the `DbWeight` from the runtime file like this, without needing to create a new trait:

before (inside migrations file):
```rust
pub type VersionCheckedMigrateV5ToV6<Runtime, Pallet, DbWeight> = VersionedRuntimeUpgrade<
	ConstU16<5>,
	ConstU16<6>,
	VersionUncheckedMigrateV5ToV6<Runtime>,
	Pallet,
	DbWeight,
>;
```

after (inside migrations file):
```rust
pub type VersionCheckedMigrateV5ToV6<Runtime, Pallet> = VersionedRuntimeUpgrade<
	ConstU16<5>,
	ConstU16<6>,
	VersionUncheckedMigrateV5ToV6<Runtime>,
	Pallet,
	<Runtime as frame_system::Config>::DbWeight,
>;
```

Then in the runtime file, you no longer need to pass the DbWeight:
```rust
parachains_configuration::migration::v6::VersionCheckedMigrateV5ToV6<
	Runtime,
	Configuration,
>,
```

Let me know if this approach is also OK, it seems to keep things clean in the runtime file without needing to create an extra trait, but I'm a noob with associated types so definitely may be missing something, please let me know if I am.